File: Baekjoon/_ETC/1717집합의표현.cpp
#include<stdio.h>
#include<string>

int findUnion(int x, int arr[]){
    if(arr[x] == x){
        return x;
    }
    arr[x] = findUnion(x,arr);

    return findUnion(arr[x],arr);
}

void makeUnion(int a, int b,int arr[]){
    int rootA,rootB;

    rootA = findUnion(a,arr);
    rootB = findUnion(a,arr);


    if(rootA < rootB){
        arr[a] = rootA;
        arr[b] = rootA;
    }
    else{
        arr[a] = rootB;
        arr[b] = rootB;
    }
    return;
}


bool isSame(int a,int b){
    if(a==b){
        return true;
    }
    else{
        return false;
    }
}

int main(){

    int n,m;
    scanf("%d %d",&n,&m);
    int arr[1000000];
    for (int i =0; i < n; i++){
        arr[i] = i + 1;
    }
    int order,a,b;
    for(int i =0; i < m; i++){
        scanf("%d %d %d", &order,&a,&b);

        if(order == 0){
            makeUnion(a,b,arr);
        }
        else if(order == 1){
            if (isSame(findUnion(a,arr),findUnion(b,arr))){
                printf("YES");
            }else{
                printf("NO");
            }
        }

    }
    return 0;

}

File: Baekjoon/UnionFind/1717.집합의표현.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

def findUnion(x):
    if arr[x] == x:
        return x
    
    arr[x] = findUnion(arr[x])

    return findUnion(arr[x])


def makeUnion(a, b):
    rootA,rootB = 0,0

    rootA = findUnion(a)
    rootB = findUnion(b)


    if rootA < rootB:
        arr[rootB] = rootA
    
    else:
        arr[rootA] = rootB

    

n, m = list(map(int,list(input().split())))
arr = [0] + [j for j in range(1,n+1)]

for i in range(m):
    order, a, b = list(map(int,list(input().split())))

    if order == 0:
        makeUnion(a,b)
    
    if order == 1:

        if findUnion(a) == findUnion(b):
            print("YES")
        else:
            print("NO")